beacon-chain/sync/validate_sync_committee_message.go

```diff
@@ -0,0 +1,278 @@
+		ctx,
+		ignoreEmptyCommittee(committeeIndices),
+		s.rejectIncorrectCommittee(committeeIndices, *msg.Topic),
+		s.ignoreHasSeenSyncMsg(m, committeeIndices),
+		s.rejectInvalidSignature(m),
+	); result != pubsub.ValidationAccept {
+		return result
+	}
+
+	s.markSyncCommitteeMessagesSeen(committeeIndices, m)
+
+	msg.ValidatorData = m
+	return pubsub.ValidationAccept
+}
+
+// Parse a sync committee message from a pubsub message.
+func (s *Service) readSyncCommitteeMessage(msg *pubsub.Message) (*ethpb.SyncCommitteeMessage, error) {
+	raw, err := s.decodePubsubMessage(msg)
+	if err != nil {
+		return nil, err
+	}
+	m, ok := raw.(*ethpb.SyncCommitteeMessage)
+	if !ok {
+		return nil, errWrongMessage
+	}
+	if m == nil {
+		return nil, errNilMessage
+	}
+	return m, nil
+}
+
+// Mark all a slot and validator index as seen for every index in a committee and subnet.
+func (s *Service) markSyncCommitteeMessagesSeen(committeeIndices []types.CommitteeIndex, m *ethpb.SyncCommitteeMessage) {
+	subCommitteeSize := params.BeaconConfig().SyncCommitteeSize / params.BeaconConfig().SyncCommitteeSubnetCount
+	for _, idx := range committeeIndices {
+		subnet := uint64(idx) / subCommitteeSize
+		s.setSeenSyncMessageIndexSlot(m.Slot, m.ValidatorIndex, subnet)
+	}
+}
+
+// Returns true if the node has received sync committee for the validator with index and slot.
+func (s *Service) hasSeenSyncMessageIndexSlot(slot types.Slot, valIndex types.ValidatorIndex, subCommitteeIndex uint64) bool {
+	s.seenSyncMessageLock.RLock()
+	defer s.seenSyncMessageLock.RUnlock()
+
+	b := append(bytesutil.Bytes32(uint64(slot)), bytesutil.Bytes32(uint64(valIndex))...)
+	b = append(b, bytesutil.Bytes32(subCommitteeIndex)...)
+	_, seen := s.seenSyncMessageCache.Get(string(b))
+	return seen
+}
+
+// Set sync committee message validator index and slot as seen.
+func (s *Service) setSeenSyncMessageIndexSlot(slot types.Slot, valIndex types.ValidatorIndex, subCommitteeIndex uint64) {
+	s.seenSyncMessageLock.Lock()
+	defer s.seenSyncMessageLock.Unlock()
+
+	b := append(bytesutil.Bytes32(uint64(slot)), bytesutil.Bytes32(uint64(valIndex))...)
+	b = append(b, bytesutil.Bytes32(subCommitteeIndex)...)
+	s.seenSyncMessageCache.Add(string(b), true)
+}
+
+// The `subnet_id` is valid for the given validator. This implies the validator is part of the broader
+// current sync committee along with the correct subcommittee.
+// We are trying to validate that whatever committee indices that were retrieved from our state for this
+// particular validator are indeed valid for this particular topic. Ex: the topic name can be
+// /eth2/b5303f2a/sync_committee_2/ssz_snappy
+// This would mean that only messages meant for subnet 2 are valid. If a validator creates this sync
+// message and broadcasts it into subnet 2, we need to make sure that whatever committee index and
+// resultant subnet that the validator has is valid for this particular topic.
+func (s *Service) rejectIncorrectCommittee(
```

Comment:
I think the naming needs to be more specific because we will likely apply such idea to attestation and beacon committee

```suggestion
func (s *Service) rejectIncorrectSyncCommittee(
```